test/python/circuit/library/test_multipliers.py

```diff
@@ -0,0 +1,123 @@
+        statevector = Statevector(circuit)
+        probabilities = statevector.probabilities()
+        pad = "0" * circuit.num_ancillas  # state of the ancillas
+
+        # compute the expected results
+        expectations = np.zeros_like(probabilities)
+        num_bits_product = num_state_qubits * 2
+        # iterate over all possible inputs
+        for x in range(2 ** num_state_qubits):
+            for y in range(2 ** num_state_qubits):
+                # compute the product
+                product = x * y % (2 ** num_result_qubits)
+                # compute correct index in statevector
+                bin_x = bin(x)[2:].zfill(num_state_qubits)
+                bin_y = bin(y)[2:].zfill(num_state_qubits)
+                bin_res = bin(product)[2:].zfill(num_bits_product)
+                bin_index = pad + bin_res + bin_y + bin_x
+                index = int(bin_index, 2)
+                expectations[index] += 1 / 2 ** (2 * num_state_qubits)
+        np.testing.assert_array_almost_equal(expectations, probabilities)
+
+    @data(
+        (3, RGQFTMultiplier),
+        (3, RGQFTMultiplier, 5),
+        (3, RGQFTMultiplier, 4),
+        (3, RGQFTMultiplier, 3),
+        (3, HRSCumulativeMultiplier),
+        (3, HRSCumulativeMultiplier, 5),
+        (3, HRSCumulativeMultiplier, 4),
+        (3, HRSCumulativeMultiplier, 3),
+        (3, HRSCumulativeMultiplier, None, CDKMRippleCarryAdder),
+        (3, HRSCumulativeMultiplier, None, DraperQFTAdder),
+        (3, HRSCumulativeMultiplier, None, VBERippleCarryAdder),
+    )
+    @unpack
+    def test_multiplication(self, num_state_qubits, multiplier, num_result_qubits=None, adder=None):
+        """Test multiplication for all implemented multipliers."""
+        if num_result_qubits is None:
+            num_result_qubits = 2 * num_state_qubits
+        if adder is not None:
+            adder = adder(num_state_qubits, kind="half")
+            multiplier = multiplier(num_state_qubits, num_result_qubits, adder=adder)
+        else:
+            multiplier = multiplier(num_state_qubits, num_result_qubits)
+        self.assertMultiplicationIsCorrect(num_state_qubits, num_result_qubits, multiplier)
+
+    @data(
+        (RGQFTMultiplier, -1),
+        (HRSCumulativeMultiplier, -1),
+        (RGQFTMultiplier, 3, 2),
+        (HRSCumulativeMultiplier, 3, 2),
+        (RGQFTMultiplier, 3, 7),
+        (HRSCumulativeMultiplier, 3, 7),
+    )
+    @unpack
+    def test_raises_on_wrong_num_bits(self, multiplier, num_state_qubits, num_result_qubits=None):
+        """Test an error is raised for a bad number of state or result qubits."""
+        with self.assertRaises(ValueError):
+            _ = multiplier(num_state_qubits, num_result_qubits)
+
+    def test_modular_cumulative_multiplier_custom_adder(self):
+        """Test an error is raised when a custom adder is used with modular cumulative multiplier."""
+        with self.assertRaises(ValueError):
```

Comment:
Is the error raised here the standard Qiskit error due to mismatching register sizes? Can we raise a more precise error here to say that the adder was the wrong size or maybe not a half-adder? 🙂

Reply:
It would be the following error that's raised:

https://github.com/Qiskit/qiskit-terra/blob/388afbd443c79923f68129c5997ec808c7bd2317/qiskit/circuit/library/arithmetic/multipliers/hrs_cumulative_multiplier.py#L95-L96

What do you think, is that understandable enough? This is more related to the fact that for modular multiplication we need varying sizes of the adder, therefore passing a custom circuit of a fixed size doesn't work. Therefore, if no adder is passed, I instantiate multiple adders of necessary sizes using the default `CDKMRippleCarryAdder`. Let me know if that sounds fine 🙂

Reply:
Yes, that sounds good 👍🏻 do you think a `NotImplementedError` or `QiskitError` might make more sense than a `ValueError`?

Reply:
`NotImplementedError` sounds a lot better, I have switched to use that, thanks for the great suggestion 🙂